### ISSUE TITLE
The Pod WatchAll will fail in about half an hour

### ISSUE THREAD

**Issue by d4ilys** [tintoy/dotnet-kube-client#157]
Hello, how does the client communicate with api-server? Is there a retry mechanism

**Comment by d4ilys** [comment]
~~~C#
  var eventStream = kubeApiClient.PodsV1()
      .WatchAll(kubeNamespace: kubeNamespace);    
	eventStream.Select(resourceEvent => resourceEvent.Resource).Subscribe(subsequentEvent =>
      {
            // do ....
      },
      error => LogError($"Listening to pod fail."),
                                                                          
      () =>
      {
          LogInfo("Listening to pod completed.");
      });
~~~

The code above will execute OnCompleted after about 0.5-1 hours

**Comment by tintoy** [comment]
Does the OnError handler ever get called? Are there any events in that time?

**Comment by tintoy** [comment]
> how does the client communicate with api-server?

The client mostly uses HTTP/HTTPS, but uses Websockets for some actions (such as Attach and Logs).

> Is there a retry mechanism

not currently but this is a good idea; we should be able to hide it behind the `IObservable` implementation.

**Comment by tintoy** [comment]
What version of Kubernetes are you using? Do you get the same behaviour when watching pods in a single namespace?

**Comment by tintoy** [comment]
Just had a look at the implementation; we just use the k8s `list` operation with `watch=true`. I suspect we may need to be a bit smarter about it, and keep track of the last event seen and resume with that event as the high water mark once the connection times out and is then re-established.

**Comment by tintoy** [comment]
Possibly related:

https://github.com/kubernetes-client/csharp/issues/533#issuecomment-1546665121

**Comment by d4ilys** [comment]
> Kubernetes version 

Kubernetes v1.23.6

> Does the OnError handler ever get called? Are there any events in that time?

No OnError events were raised, and the connection should be disconnected normally

If the connection timeout is caused, an exception message is caught

After about 40-50 minutes, the OnCompleted will be triggered

> Temporary solution

```c#
void InternalWatch()
{
    var eventStream = kubeApiClient.PodsV1()
        .WatchAll(kubeNamespace: kubeNamespace);
    eventStream.Select(resourceEvent => resourceEvent.Resource).Subscribe(subsequentEvent =>
        {
          
            // do ...
        },
        error => LogError($"Listening to pod fail."),
        () =>
        {
            LogInfo("Listening to pod completed.");
            // Retrigger Watch
            InternalWatch();
        });
}
```

Do you plan to repeat or fix this problem in the next step? Thank you for your positive reply

**Comment by tintoy** [comment]
Thanks, that confirms my initial impression. I’ll look into adding a flag argument to automatically reconnect (since it’s a change in behaviour I’ll be making it opt-in). 🙂

**Comment by tintoy** [comment]
Could you do me a favor and call `.ToString()` on the error callback (where you do `LogError`). It would be helpful to see the exception type and stack trace.

**Comment by d4ilys** [comment]
This is the scenario I was debugging locally, and nothing unusual happened.

**![f55b86ff01725cadd642d547b9c0ef9](https://github.com/tintoy/dotnet-kube-client/assets/54463101/6c74bf99-cdfb-411e-b310-e109798690f4)
![4f414d2f11cf2481fd6491b4aa3d7b7](https://github.com/tintoy/dotnet-kube-client/assets/54463101/74b59e09-8a02-4f1f-8da6-501cc84966bc)**

**Comment by tintoy** [comment]
This will require targeting `netstandard2.0` instead of `netstandard1.4`, but I don't think that's likely to be a problem for consumers given how old that was.

**Comment by tintoy** [comment]
If you have time, would you be able to try out the latest development version (`2.5.9-develop.3`) of the package?

Development package feed is at https://www.myget.org/F/dotnet-kube-client/api/v3/index.json

**Comment by tintoy** [comment]
I’ll be publishing these changes to NuGet in a day or 2.

**Comment by raman-m** [comment]
@d4ilys As the user who raised the issue, could you please confirm if the bug fix from #159 has been applied?

**Comment by tintoy** [comment]
Fixed in [v2.5.10](https://www.nuget.org/packages/KubeClient/2.5.10).
